### ISSUE TITLE
Fix false positives for escaped multi-line in `string-no-newline`

### ISSUE THREAD

**Issue by AlexWayfer** [stylelint/stylelint#4057]
> Clearly describe the bug

I want to split large strings for aesthetic reasons with enabled `string-no-newline` rule.

There is the quote from [the specs](https://www.w3.org/TR/CSS2/syndata.html#strings), including:

> And also: "It is possible to break strings over several lines, for aesthetic or other reasons, but in such a case the newline itself has to be escaped with a backslash (\)."

But for some reason escaped newlines are forbidden in Stylelint.

> Which rule, if any, is the bug related to?

`string-no-newline`

> What CSS is needed to reproduce the bug?

```css
background-image: url(
  "data:image/svg+xml;utf8,\
  <svg fill='black' height='24' \
    viewBox='0 0 24 24' width='24' \
    xmlns='http://www.w3.org/2000/svg'\
  ><path d='M7 9l5 5 5-5z'/></svg>"
);
```

> What stylelint configuration is needed to reproduce the bug?

```yaml
extends: stylelint-config-standard
```

> Which version of stylelint are you using?

`9.10.1`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI with `stylelint assets/styles/`.

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

No.

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged:

```shell
 116:197  ✖  Expected line length to be no more than 80 characters                                               max-line-length
```

**Comment by hudochenkov** [comment]
Thanks for the report and for using the template.

I can't reproduce this error with the following code:

```css
.hello {
  background-image: url(
    "data:image/svg+xml;utf8,\
    <svg fill='black' height='24' \
      viewBox='0 0 24 24' width='24' \
      xmlns='http://www.w3.org/2000/svg'\
    ><path d='M7 9l5 5 5-5z'/></svg>"
  );
}
```

Looks like you're also has `max-line-length: 80` in your config. I can't reproduce error with this rule enabled as well.

Could you check please what is actual CSS, which triggers warning? Maybe there are many indentation symbols, and line is actually bigger than 80 symbols.

**Comment by hudochenkov** [comment]
Looks like our demo is misleading, it doesn't show problem, when CLI shows.

For the following CSS:

```css
.hello {
  background-image: url("data:image/svg+xml;utf8,\
	  <svg fill='black' height='24' \
		viewBox='0 0 24 24' width='24' \
		xmlns='http://www.w3.org/2000/svg'\
	  ><path d='M7 9l5 5 5-5z'/></svg>");
}
```

I got this violation:

```
2:51  ✖  Unexpected newline in string   string-no-newline
```

Looks like it's a bug.

As the issue has been labelled "help wanted", please consider [contributing a fix](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fixing-a-bug-in-an-existing-rule).

P. S. In original issue violation is incorrect, which confused me.

**Comment by AlexWayfer** [comment]
> P. S. In original issue violation is incorrect, which confused me.

Yes, my bad, sorry. There was a mess with #4059. But there're still two problems.

**Comment by vankop** [comment]
@hudochenkov Same problem #4059, postcss does not parse data url node:
Steps to reproduce:

1. __lib/rules/max-line-length/tests/index.js__
```
"use strict";

const rule = require("..");
const { messages, ruleName } = rule;

const testUrl = `data:image/svg+xml;utf8, \
      <svg fill='black' height='24' \
        viewBox='0 0 24 24' width='24' \
        xmlns='http://www.w3.org/2000/svg'\
      ><path d='M7 9l5 5 5-5z'/></svg>`;

testRule(rule, {
  ruleName,
  config: [true],

  accept: [
    {
      code: `div { background: "${testUrl}"}`
    },
    {
      code: `div { background: "${encodeURI(testUrl)}"}`
    },
]});
```

2. run `node --inspect-brk node_modules/jest/bin/jest.js --runInBand lib/rules/string-no-newline/__tests__/index.js`

3. In debugger add breakpoint/`debugger;` here
https://github.com/stylelint/stylelint/blob/master/lib/getPostcssResult.js#L134
In first run `result.result.root.nodes: []`, in second run `result.result.root.nodes: [Node]`

**Comment by vankop** [comment]
> Looks like our demo is misleading, it doesn't show problem, when CLI shows.

For CLI it parses 🙈, but failing after on `https://github.com/stylelint/stylelint/blob/master/lib/rules/string-no-newline/index.js#L83` parser returns source string instead of removing `\` and newline symbols

**Comment by Mouvedia** [comment]
related: postcss/postcss#1349

**Comment by rgpublic** [comment]
I wonder why under postcss/postcss#1349 @romainmenke  writes:

> Took me a while to analyse this issue because initially I interpreted this as "PostCSS has a bug with escaped symbols".
But this is not a bug in PostCSS. As far as I can tell PostCSS produces the correct AST with or without escaped symbols.
It produces a few words too many at the tokenizer level, but this goes away in the parser.

If I understood @vankop (point 3.) correctly, we don't get any nodes for this string. According to the above statement we should get at least sth. so this could be dealt with within stylelint.

Which one is correct? Are we getting sth. we can detect and deal with within stylelint? Or is it currently impossible to solve without PostCSS fixing this first? Wouldn't this also be a problem for PostCSS as this is (AFAIK) valid CSS?

And - final question ;-) - shouldn't the documentation be updated for the time being, because currently it just isn't working and causes (at least for me) a lot of confusion? Currently, the stylelint docs explicitly state that this should work.

**Comment by romainmenke** [comment]
This definitely is not a PostCSS issue :)
I checked it multiple times.

------

I wasn't aware of this issue otherwise I would have submitted a patch here earlier.
Thank you for pinging me here @rgpublic 

------

Worth noting that the tests for this aspect are currently incorrect: 
https://github.com/stylelint/stylelint/blob/cb3a7f8d5e89f2e467a4166293c17d3c2d1e72cc/lib/rules/string-no-newline/__tests__/index.mjs#L16-L18

This only has `\\n` which is an escaped newline.
What it should be instead is `\\\n` -> an escaped backslash and a newline.

Because the test is incorrect the actual rule also has bugs because it is testing the wrong things.

------

Luckily this is something that I have already fixed by writing a spec compliant tokenizer 😄 
The spec has a very specific branch and error for this case.

So we can simply passthrough the tokenizer error to the stylelint report.

I'll submit a patch later today.

**Comment by Gitoffthelawn** [comment]
There is a regression in the latest release (v1.5.48) of the [Stylus web browser extension](https://github.com/openstyles/stylus), most likely due to Stylus updating Stylelint to v16.5.0.

When Stylelint's `string-no-newline` is enabled, the following valid CSS code is now flagged as an error:

```
	.test {
		mask:
			url(
				'https://www.example.com/foo/bar/baz\
					/svg/exampleMask.svg'
			) no-repeat 2px 0;
	}
```

Will the patch submitted by @romainmenke fix this issue (if/when Stylus includes that patch)?

**Comment by romainmenke** [comment]
I couldn't say @Gitoffthelawn without a reproduction :)

If the issue is still present after everything has been released and updated you should open a new issue, then we can investigate further.

**Comment by Gitoffthelawn** [comment]
@romainmenke Thanks.  Will do.  BTW, can you quickly run the above code snippet I wrote through your Stylelint test build that includes your PR to see if it generates a warning or error?

Also, if anyone interfaces with a repo project manager at Stylus (perhaps @tophf, @eight04, or @Mottie?), this GitHub account cannot post issues or PRs in that repo, so perhaps someone there can look into making it so it can.

**Comment by tophf** [comment]
When reporting stylelint issues please use the [playground](https://stylelint.io/demo/) to reproduce the issue with your config. If it's not reproducible then it's not a problem in stylelint, but rather in the extension.

**Comment by Gitoffthelawn** [comment]
@tophf Thanks. Is that playground working correctly in Firefox?  If so, is there documentation for it?  I ask because whatever I enter in the left pane does not affect the right pane (except for adding plus signs before changed lines), even after I press the arrow button between the two panes.  Also, the "Problems" tab always indicates *"No problems!"*, even if I intentionally enter invalid CSS syntax in the left pane for testing purposes.

**Comment by tophf** [comment]
It works in Firefox, here's the output when using default settings:

![image](https://github.com/stylelint/stylelint/assets/1310400/73f456d4-e6b9-41df-9243-ff1d270e3105)

**Comment by Gitoffthelawn** [comment]
@tophf Weird. Did you simply paste the test code I wrote into the left pane of the playground?  This is what I get when I do that:  
![stylelint io playground](https://github.com/stylelint/stylelint/assets/7672848/01ff9ada-3b15-4749-a193-4f5cf6504e8f)

**Comment by eight04** [comment]
There is also a playground at openstyles/stylelint-bundle:
https://raw.githack.com/openstyles/stylelint-bundle/master/demo/index.html

**Comment by tophf** [comment]
@Gitoffthelawn, try disabling all ad-blocking or script-blocking extensions.

**Comment by Gitoffthelawn** [comment]
@tophf Thanks. That didn't do it... debugging playground's code now to determine what is broken.  I'll file an issue report once I determine the cause.

**Comment by Gitoffthelawn** [comment]
Found the issue in the playground.  Submitted issue report: https://github.com/stylelint/stylelint-demo/issues/412

**Comment by Gitoffthelawn** [comment]
@eight04 Thanks for the link to the additional playground.

**Comment by Gitoffthelawn** [comment]
Both playgrounds exhibit the Stylelint bug when checking the test CSS code I wrote, but I'll wait to hear/see if the PR that @romainmenke wrote fixes the bug before filing another issue report here (unless someone wants me to, in which case I can do so).

**Comment by Gitoffthelawn** [comment]
@tophf Sorry to remind here, but I don't have an email address for you. Can you please add access to this GitHub account so I can post Stylus issues and PRs on https://github.com/openstyles/stylus? I have been doing so from other GitHub accounts, but I prefer to work on front-end dev issues from this account. Thank you.

**Comment by tophf** [comment]
I don't have access to settings of the repo. P.S. You can see my email yourself on github by opening any recent commit I've made with `.patch` added to the URL, which will show the full info.

**Comment by Gitoffthelawn** [comment]
@tophf Thanks.  I forgot about that trick (although it feels a bit like snooping, as probably not everyone knows it's that simple).  Who currently has access to the repo's settings?

**Comment by tophf** [comment]
1) It's an integral part of git itself, so anyone who uses git should know it. 2) We shouldn't hijack this thread.

**Comment by romainmenke** [comment]
@Gitoffthelawn please refrain making this thread longer with off topic questions :)
There are a lot of people in this thread and they all get notifications and/or emails each time you post something. This causes a lot of noise.
